PatelDLTGroupStamps/Views/StampStockView.swift

```diff
@@ -8,27 +8,79 @@
 import SwiftUI
 
 struct StampStockView: View {
+
+  @StateObject private var stampHelper = StampHelper()
+  @State private var shoppingAlert = false
   
   var body: some View {
     
     // MARK: Assignment 2: Summary screen and detail screen
-    
     NavigationStack {
-      List {
-        ForEach(stamp.unitedNationsStamps, id: \.self) {
-          unitedNationStamp in
-          NavigationLink(unitedNationStamp.description, destination: StampView(stamp: unitedNationStamp))
+      List(stampHelper.stamps, id: \.id) { stamp in
+        Spacer()
+        AsyncImage(url: URL(string: stamp.image)!) { stamp in
+          if let image = stamp.image {
+            image
+              .resizable()
+              .aspectRatio(contentMode: .fit)
+            // MARK: Assignment 3
+          } else if stamp.error != nil {
+            Button("Show Alert") {
+              shoppingAlert = true
+            }
+            Text("No image avaiable")
+          } else {
+            Image(systemName: "photo")
+              .resizable()
+              .aspectRatio(contentMode: .fill)
+          }
+        }
+        Text(stamp.itemId)
+        Text(stamp.title)
+        Text(stamp.price)
+        Text(stamp.categoryPath)
+        Text(stamp.categoryIdPath)
+          .multilineTextAlignment(.leading)
+          .font(.headline)
+      }
+      .navigationTitle("Stocked Stamps")
+      .listStyle(.grouped)
+      .navigationBarItems(trailing:
+                            Button(action: {
+        Task {
+          try await stampHelper.saveStamps()
         }
-        .navigationTitle("Stocked Stamps")
+      }) {
+        Image(systemName: "plus")
+      })
+    }
+    // MARK: Assignment 3 - Alert setup
+    .alert("No network connection", isPresented: $shoppingAlert) {
+      Button("OK", role: .cancel) {}
+    }
+    // MARK: Assignment 2 Continued - Custom Error Setup
+    .task {
+      do {
+        try await stampHelper.getStamps()
+      } catch StampError.badURLRequest {
+        print("bad url request")
```

Comment:
I know we're in the thick of dev work, but it's worth the extra time to write your error messages the right way the first time around. This is tech debt you'll now have to handle before you submit your capstone 🙃